scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulerFrontend.java

```diff
@@ -1915,20 +1918,37 @@ public void endActivity(Body body) {
 
     @Override
     @ImmediateService
-    public boolean addJobSignal(String sessionId, String jobId, String signal) {
+    public Set<String> addJobSignal(String sessionId, String jobId, String signal) throws SignalApiException {
         try {
 
-            Set<String> signals = (Set) publicStore.get(SIGNAL_ORIGINATOR, SIGNAL_TASK_ID, signalsChannel, jobId);
+            Set<String> signals = (HashSet<String>) publicStore.get(SIGNAL_ORIGINATOR,
+                                                                    SIGNAL_TASK_ID,
+                                                                    signalsChannel,
+                                                                    jobId);
             if (signals == null) {
                 signals = new HashSet<>();
             }
 
+            // Throw an error if: (i) the corresponding ready signal does not exist and (ii) the signal is not a ready signal
+            String readyPrefix = SignalApiImpl.READY_PREFIX;
+            if (!signals.contains(readyPrefix + signal) && !signal.startsWith(readyPrefix)) {
+                throw new SignalApiException("Job " + jobId + " is not ready to receive the signal " + signal);
+            }
+
```

Comment:
to improve atomicity you can do:
` Set<String> signals = (HashSet<String>) publicStore.get(SIGNAL_ORIGINATOR,
                                                                    SIGNAL_TASK_ID,
                                                                    signalsChannel,
                                                                    jobId);
if (!signals.contains(readyPrefix + signal) && !signal.startsWith(readyPrefix)) {
                throw new SignalApiException("Job " + jobId + " is not ready to receive the signal " + signal);
}                                                                    
`
return publicStore.compute(SIGNAL_ORIGINATOR, SIGNAL_TASK_ID, signalsChannel, jobId, "{ k, x ->x.remove('" + READY_PREFIX + signalName + "'); x.add('" + signalName + "'); x}");

Reply:
additionnally, is it meaningful to allow sending a ready signal, like in the test signal.startsWith(readyPrefix) ?